### ISSUE TITLE
DynamicServerError at build time when cache: no-store is enabled

### ISSUE THREAD

**Issue by mpotane** [apollographql/apollo-client-nextjs#339]
```ts
export const { query } = registerApolloClient(() => {
  return new ApolloClient({
    cache: new InMemoryCache(),
    link: new HttpLink({
      uri: "https://api.github.com/graphql",
      // you can disable result caching here if you want to
      // (this does not work if you are rendering your page with `export const dynamic = "force-static"`)
      fetchOptions: { cache: "no-store" }, // --enabled
      headers: {
        authorization: `Bearer ${process.env.GITHUB_TOKEN}`,
      },
    }),
  });
});
```

**Comment by phryneas** [comment]
Hi Mark,
could you please add a little bit more explanation?

Generally, if you force your page to be generated statically and then make a dynamic request like this, Next.js will error, as these two modes cannot exist at the same time.

I assume that's what's happening here - but with more context I can't help you.

**Comment by phryneas** [comment]
@Tagir-A I just found this issue: https://github.com/vercel/next.js/issues/46737#issuecomment-2449603499

> If you catch errors using try-catch, let errors with error.digest === 'DYNAMIC_SERVER_USAGE' pass through unhandled, then everything will work without problems:

It seems that Next.js started using this error to detect dynamic usage. I'm not sure if we can prevent Apollo Client from actually throwing these - I'll look into it.

**Comment by phryneas** [comment]
I tried reproducing this and I don't see it - you are using `export const dynamic = "force-dynamic";`, too? That's not enabled in Next by default.

**Comment by phryneas** [comment]
You can look at my reproduction here: https://github.com/phryneas/repro_next_dynamic-server-error

Did you do anything differently?

**Comment by phryneas** [comment]
The only way I can see this happening is if I remove the `await` from the `query` call, but that is [documented Next.js behaviour](https://nextjs.org/docs/messages/dynamic-server-error)

**Comment by Tagir-A** [comment]
@phryneas 

> I tried reproducing this and I don't see it - you are using export const dynamic = "force-dynamic";, too? That's not enabled in Next by default.

Nope, I tried to opt-out of caching on [data level](https://nextjs.org/docs/14/app/building-your-application/data-fetching/fetching-caching-and-revalidating#opting-out-of-data-caching) by using `cache: "no-store"`

```
    const { data, error, loading } = await getApolloClientOnServer().query<{
        content: {
          id
        };
    }>({
        query: QUERY,
        context: {
            fetchOptions: {
                // This should work, but produces an error "ApolloError: Dynamic server usage: no-store fetch"
                cache: 'no-store',
            },
        },
    });
```

So as a workaround, I can opt-out on route level:
```
// ...

export const dynamic = 'force-dynamic';

export default function Page() {
// ...
}
```

This works, but usually I need to opt-out of caching on specific requests, rather than on specific routes.
Cause the route usually contains lots of logic, and if all components which load dynamic data are moved out of the route (due to refactor or user flow change), I want it automatically opt-in to caching. This would be possible if caching is disabled on data level, but it's not possible when it's disabled on route level.

**Comment by phryneas** [comment]
I'm honestly unsure at this point if Next.js 15 even supports that anymore - my example was really minimal and it was already failing there :/
